thrust/thrust/mr/pool.h

```diff
@@ -451,35 +476,47 @@ class unsynchronized_pool_resource final
             );
 
             oversized_block_descriptor desc = *block;
+            assert(desc.current_size == n);
+            assert(desc.alignment == alignment);
 
             if (m_options.cache_oversized)
             {
                 desc.next_cached = m_cached_oversized;
-                *block = desc;
+
+                if (desc.size != n) {
+                    desc.current_size = desc.size;
+                    block = static_cast<oversized_block_descriptor_ptr>(
+                        static_cast<void_ptr>(static_cast<char_ptr>(p) +
+                                              desc.size));
+                    if (detail::pointer_traits<
+                            oversized_block_descriptor_ptr>::get(desc.prev)) {
+                        thrust::raw_reference_cast(*desc.prev).next = block;
+                    } else {
+                        m_oversized = block;
+                    }
+
+                    if (detail::pointer_traits<
+                            oversized_block_descriptor_ptr>::get(desc.next)) {
+                        thrust::raw_reference_cast(*desc.next).prev = block;
+                    }
+                }
+
                 m_cached_oversized = block;
+                *block = desc;
 
                 return;
             }
 
-            if (!detail::pointer_traits<oversized_block_descriptor_ptr>::get(desc.prev))
-            {
-                assert(m_oversized == block);
+            if (detail::pointer_traits<oversized_block_descriptor_ptr>::get(
```

Comment:
Maybe even pull the alias up so that we can reuse it here or make it a member of the class?